### PR TITLE
Fix automatic conversion of Auth to Realm URLs

### PR DESCRIPTION
**Pull request by cmelchior**
Closes https://github.com/realm/realm-js/issues/1980
Closes https://github.com/realm/realm-js/issues/1981

Fixes how we do automatic conversions of Auth URLs to Realm URLs. This also adds support for defining URL's as relative to the auth host.

### CODE REVIEW

File: lib/user-methods.js [hunk 1]
```diff
@@ -62,6 +62,18 @@ function checkObjectTypes(obj, types) {
     }
 }
 
+function normalizeSyncUrl(authUrl, syncUrl) {
+    let parsedAuthUrl = new URL(authUrl);
+    let secure = (parsedAuthUrl.protocol === 'https:')?'s':'';
```
Comment (bmunkholm) [inline lib/user-methods.js]:
```suggestion
    let secure = (parsedAuthUrl.protocol === 'https:') ? 's' : '';
```

File: lib/user-methods.js [hunk 2]
```diff
@@ -629,18 +641,16 @@ const instanceMethods = {
...
-        }
 
-        // Create default config
-        let url = new URL(this.server);
-        let secure = (url.protocol === 'https:')?'s':'';
-        let port = (url.port === undefined)?'9080':url.port;
-        let realmUrl = `realm${secure}://${url.hostname}:${port}/default`;
+            if (config.sync.url) {
+                config.sync.url = new URL(config.sync.url, {})
+
```
Comment (bmunkholm) [inline lib/user-methods.js]:
```suggestion
```

**Comment by cmelchior** [comment]
Ready for review.
@kraenhansen I'm not missing any use cases from you, right?

File: lib/user-methods.js [hunk 3]
```diff
@@ -62,6 +62,19 @@ function checkObjectTypes(obj, types) {
...
+function normalizeSyncUrl(authUrl, syncUrl) {
+    let parsedAuthUrl = new URL(authUrl);
+    let secure = (parsedAuthUrl.protocol === 'https:') ? 's' : '';
+    // Inherit ports from the Auth url
+    let port = (parsedAuthUrl.port === undefined) ? "" : `:${parsedAuthUrl.port}`;
+    let baseUrl = `realm${secure}://${parsedAuthUrl.hostname}${port}`;
+    if (syncUrl) {
+        return new URL(syncUrl, baseUrl, false).toString();
+    } else {
+        return baseUrl + "/default";
```
Comment (nirinchev) [inline lib/user-methods.js]: You can avoid the multiple returns if you supply a default value to the `syncUrl`. Something like:

```js
if (!syncUrl) {
    syncUrl = "/default";
}

return new URL(syncUrl, baseUrl, false).toString();
```

File: lib/user-methods.js [hunk 4]
```diff
@@ -62,6 +62,19 @@ function checkObjectTypes(obj, types) {
     }
 }
 
+function normalizeSyncUrl(authUrl, syncUrl) {
+    let parsedAuthUrl = new URL(authUrl);
+    let secure = (parsedAuthUrl.protocol === 'https:') ? 's' : '';
```
Comment (nirinchev) [inline lib/user-methods.js]: This seems like an overly clever way to save a few letters 😛I think it'd be much more obvious if we were explicit about what this is and why we need it. Something like:

```js
const realmProtocol = parsedAuthUrl.protocol === "https:" ? "realms" : "realm";
```

File: lib/user-methods.js [hunk 5]
```diff
@@ -62,6 +62,19 @@ function checkObjectTypes(obj, types) {
     }
 }
 
+function normalizeSyncUrl(authUrl, syncUrl) {
+    let parsedAuthUrl = new URL(authUrl);
```
Comment (nirinchev) [inline lib/user-methods.js]: These all should be `const` (although I know that throughout the codebase we have been very liberal with the usage of `let`).

File: lib/user-methods.js [hunk 6]
```diff
@@ -62,6 +62,19 @@ function checkObjectTypes(obj, types) {
...
+    let secure = (parsedAuthUrl.protocol === 'https:') ? 's' : '';
+    // Inherit ports from the Auth url
+    let port = (parsedAuthUrl.port === undefined) ? "" : `:${parsedAuthUrl.port}`;
```
Comment (nirinchev) [inline lib/user-methods.js]: You can safely check the port for falsy-ness and save a few characters:

```suggestion
    let port = parsedAuthUrl.port ? `:${parsedAuthUrl.port}` : "";
```

File: lib/user-methods.js [hunk 7]
```diff
@@ -629,18 +642,16 @@ const instanceMethods = {
...
             }
-        }
 
-        // Create default config
-        let url = new URL(this.server);
-        let secure = (url.protocol === 'https:')?'s':'';
-        let port = (url.port === undefined)?'9080':url.port;
-        let realmUrl = `realm${secure}://${url.hostname}:${port}/default`;
+            if (config.sync.url) {
+                config.sync.url = new URL(config.sync.url, {})
```
Comment (nirinchev) [inline lib/user-methods.js]: Why do we need to change the type here?

**Comment by kraenhansen** [comment]
@cmelchior I was looking for how to use this in the documentation https://realm.io/docs/javascript/2.22.0/api/Realm.Sync.User.html. I'm afraid we forgot to update the docs with this change in behaviour.

**Pull request merged**